Storage/SQL/BrowserTable.swift

```diff
@@ -240,31 +268,57 @@ public class BrowserTable: Table {
             return drop(db) && create(db, version: to)
         }
 
-        log.debug("Updating browser tables from \(from) to \(to).")
-        if from < 4 {
+        if from > to {
+            // This is likely an upgrade from before Bug 1160399.
+            log.debug("Downgrading browser tables. Assuming drop and recreate.")
             return drop(db) && create(db, version: to)
         }
 
-        if from < 5 {
-            let queue =
-            "CREATE TABLE IF NOT EXISTS \(TableQueuedTabs) (" +
-            "url TEXT NOT NULL UNIQUE, " +
-            "title TEXT" +
-            ") "
-            if !self.run(db, queries: [queue]) {
+        if from < 4 && to >= 4 {
+            return drop(db) && create(db, version: to)
+        }
+
+        if from < 5 && to >= 5  {
+            let queries: [(String?, Args?)] = [(getQueueTableCreationString(forVersion: to), nil)]
+            if !self.run(db, queries: queries.filter({ $0.0 != nil }).map({ ($0.0!, $0.1) })) {
                 return false
             }
         }
 
-        if from < 6 {
-            let queries = [
-                "DROP INDEX IF EXISTS \(IndexVisitsSiteIDDate)",
-                "CREATE INDEX IF NOT EXISTS \(IndexVisitsSiteIDIsLocalDate) ON \(TableVisits) (siteID, is_local, date)",
+        if from < 6 && to >= 6 {
+            if !self.run(db, queries: [
+                ("DROP INDEX IF EXISTS \(IndexVisitsSiteIDDate)", nil),
+                ("CREATE INDEX IF NOT EXISTS \(IndexVisitsSiteIDIsLocalDate) ON \(TableVisits) (siteID, is_local, date)", nil)
+            ]) {
+                return false
+            }
+        }
+
+        if from < 7 && to >= 7 {
+            let queries: [(String?, Args?)] = [
+                (getDomainsTableCreationString(forVersion: to), nil),
+                ("ALTER TABLE \(TableHistory) ADD COLUMN domain_id INTEGER REFERENCES \(TableDomains)(id) ON DELETE CASCADE", nil)
             ]
 
-            if !self.run(db, queries: queries) {
+            if !self.run(db, queries: queries.filter({ $0.0 != nil }).map({ ($0.0!, $0.1) })) {
                 return false
             }
+
+            let cursor = db.executeQuery("SELECT * FROM \(TableHistory)", factory: { (row) -> (id: Int, url: String) in
+                return (row["id"] as! Int, row["url"] as! String)
+            }, withArgs: nil)
+
+            for row in cursor {
```

Comment:
This might be five thousand history rows, resulting in ten thousand queries being run on database upgrade. I think we need to do better; even at 1msec per query that's ten seconds.